start_admin_sshd.sh

```diff
@@ -6,38 +6,87 @@ set -e
 
 mkdir -p /home/ec2-user/.ssh/
 chmod 700 /home/ec2-user/.ssh/
-ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
-ssh_config_dir="/home/ec2-user/.ssh"
-
-# Populate authorized_keys with all the public keys found in instance meta-data
-# The URLs for keys include an index and the keypair name, e.g.
-# http://169.254.169.254/latest/meta-data/public-keys/0=mykeypair/openssh-key
-ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
-touch ${ssh_authorized_keys}
-chmod 600 ${ssh_authorized_keys}
-public_key_base_url="http://169.254.169.254/latest/meta-data/public-keys/"
-imds_session_token=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 60")
-imds_request_add_header="X-aws-ec2-metadata-token: ${imds_session_token}"
-public_key_indexes=($(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}" \
-    | cut -d= -f1 \
-    | xargs))
-
-for public_key_index in "${public_key_indexes[@]}"; do
-  public_key_data="$(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}/${public_key_index}/openssh-key")"
-  if [[ ! "${public_key_data}" =~ ^"ssh" ]]; then
-    echo "Key ${public_key_data} with index ${public_key_index} looks invalid" >&2
-    continue
+declare -r ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
+declare -r ssh_config_dir="/home/ec2-user/.ssh"
+declare -r sshd_config="/etc/ssh/sshd_config"
+declare -r user_data="/.bottlerocket/host-containers/admin/user-data"
+available_auth_methods=0
+
+OLDIFS="${IFS}"
+IFS=$'\n'
+
+# Is user-data does not contain a JSON block, there's not much point in continuing.
+if [[ ! $(jq empty "${user_data}" &>/dev/null) ]]; then
+  echo "${user_data} does not contain valid JSON." >&2
+fi
+
+# Populate authorized_keys with all the public keys found in api user-data
+if grep -q authorized_keys "${user_data}"; then
```

Comment:
I'm not sure that the "invalid json check" will trigger, unless the file doesn't exist:

```bash
jq empty /dev/null ; echo $?
# prints: 0
```

This might be due to a difference in the version of jq used, but I'm thinking that it might be irrelevant. Instead, the follow up check can check for this.

```bash
if ! jq -e '.["ssh"]["authorized_keys"] as $val | ($val | length > 0)' <<< "$json_data"
then
  log "ERROR: user data does not specify ssh.authorized_keys"
else
  # handle ssh-authorized keys
fi
```
<details>
```
( for key in value_key array_key empty_key empty_array null_key; do ( set -x; jq -e --arg key $key '.[$key]? as $val | ($val | length > 0)' <<< '{"value_key": "imavalue", "array_key": [42], "empty_key": "", "empty_array": [], "null_key": null}' ; ); echo $?; done )
+ jq -e --arg key value_key '.[$key]? as $val | ($val | length > 0)'
true
0
+ jq -e --arg key array_key '.[$key]? as $val | ($val | length > 0)'
true
0
+ jq -e --arg key empty_key '.[$key]? as $val | ($val | length > 0)'
false
1
+ jq -e --arg key empty_array '.[$key]? as $val | ($val | length > 0)'
false
1
+ jq -e --arg key null_key '.[$key]? as $val | ($val | length > 0)'
false
1
```
</details>

In addition, lifting the procedure into a function will allow a cleaner short circuiting.

```bash
process_ssh_authorized_keys() {
    # allow injected values for testing
	local json_input="${1:-/dev/stdin}"
    if ! jq -e '.["ssh"]["authorized_keys"] as $val | ($val | length > 0)' < "$json_input"
    then
        log "ERROR: user data does not specify ssh.authorized_keys"
        return 1
    fi

    # handle ssh-authorized keys
    
    # print collected keys for installation
    ( IFS=$'\n'; echo "${validated_keys[@]}" )
}

# ... in script ...
if authorized_keys=$(process_authorized_keys); then
    install -D /dev/null -m 0600 -o ec2-user "$ec2_user/.ssh/authorized_keys"
    echo "$authorized_keys" > "$ec2_user/.ssh/authorized_keys"
    let auth_available++
fi
```

Reply:
jq empty seemed to work in my testing, but I'll admit it's always felt a little hacky. This seems more robust.